lib/oli/google_docs/markdown/renderer.ex

```diff
@@ -0,0 +1,508 @@
+
+        inline_paragraph_group?(children) ->
+          children
+          |> Enum.map(&render_paragraph_segment/1)
+          |> Enum.reject(&(&1 == ""))
+          |> join_segments_preserving_space()
+          |> restore_placeholders()
+
+        true ->
+          children
+          |> Enum.flat_map(&render_cell_child/1)
+          |> Enum.reject(&(&1 == ""))
+          |> Enum.join("\n\n")
+          |> restore_placeholders()
+      end
+
+    result
+    |> normalize_block_latex()
+    |> String.trim()
+  end
+
+  defp render_cell_child(text) when is_binary(text) do
+    content = restore_placeholders(text)
+
+    if String.trim(content) == "" do
+      []
+    else
+      [content]
+    end
+  end
+
+  defp render_cell_child({"p", _attrs, children, _meta}) do
+    case render_inline(children) do
+      "" -> []
+      inline -> [inline]
+    end
+  end
+
+  defp render_cell_child({"ul", _attrs, items, _meta}) do
+    [render_list(items, fn _index -> "- " end)]
+  end
+
+  defp render_cell_child({"ol", _attrs, items, _meta}) do
+    [render_list(items, fn index -> Integer.to_string(index) <> ". " end)]
+  end
+
+  defp render_cell_child({"blockquote", _attrs, children, _meta}) do
+    quoteds =
+      children
+      |> Enum.flat_map(&render_cell_child/1)
+      |> Enum.flat_map(&String.split(&1, "\n"))
+      |> Enum.map(&("> " <> &1))
+
+    case quoteds do
+      [] -> []
+      _ -> [Enum.join(quoteds, "\n")]
+    end
+  end
+
+  defp render_cell_child(
+         {"pre", _attrs, [{"code", code_attrs, code_children, _meta_code}], _meta}
+       ) do
+    language =
+      code_attrs
+      |> Enum.into(%{})
+      |> Map.get("class", "")
+      |> String.replace_prefix("language-", "")
+      |> String.trim()
+
+    code = Enum.map(code_children, &render_plain_text/1) |> Enum.join()
+    fence = "```"
+    lang_segment = if language == "", do: "", else: language
+    [Enum.join([fence <> lang_segment, code, fence], "\n")]
+  end
+
+  defp render_cell_child({"img", attrs, _children, _meta}) do
+    attrs_map = Enum.into(attrs, %{})
+    alt = attrs_map["alt"] || ""
+    title = attrs_map["title"]
+    src = attrs_map["src"] || ""
+
+    image = ["![", alt, "](", src, render_image_title(title), ")"] |> IO.iodata_to_binary()
+
+    case String.trim(image) do
+      "" -> []
+      content -> [content]
+    end
+  end
+
+  defp render_cell_child({"a", attrs, [{"img", _img_attrs, _children, _} = img], _meta}) do
+    img_attrs = Enum.into(attrs, %{})
+    href = img_attrs["href"] || ""
+
+    render_cell_child(img)
+    |> Enum.map(fn image -> "[" <> image <> "](" <> href <> ")" end)
+  end
+
+  defp render_cell_child({tag, attrs, children, meta}) when tag in ["div", "section", "span"] do
+    render_cell_child({"p", attrs, children, meta})
+  end
+
+  defp render_cell_child({tag, _attrs, children, _meta})
+       when tag in ["tbody", "thead", "tr", "td"] do
+    children
+    |> Enum.flat_map(&render_cell_child/1)
+  end
```

Comment:
This block makes no difference so it could be removed and we could keep only the one below.